proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+# Problem Statement
+
+Cryptography in Haskell lacks significant capability beyond basic primitives. This places a significant burden on developers to properly implement various security techniques, and exposes end users to significant risk in the event of a lapse in security. Some companies have built their own solution to this, but there is no community-driven, community-owned solution.
+
+Cryptography in Haskell is also fragile and outdated, having seen considerable flux / churn over the years as various libraries have been developed and then deprecated or abandoned in favor of newer ones. This has placed the long-term stability of important libraries such as `tls` and `x509` at risk of falling behind as compilers are upgraded, old standards are updated, and new standards are implemented.
+
+There are several reasons for this:
+
+1) Cryptography involves a lot of stateful low-level bit twiddling and random IO, something Haskell is not known for being good at. Classes like `Bits` and `FiniteBits` classes are critical for cryptography, but are terribly awkward to implement (like `Num`). Also, `ByteString` does not implement them for nuanced reasons.
+
+2) Cryptography has a high knowledge and technical skill floor. Cryptography is more than just the usual primitives of ciphers, public and private keys, hashes, and MACs, and the scope of knowledge can grow quickly as primitives are combined into more advanced algorithms.
+
+3) The notion that functional programming is "bad at cryptography" (ie, underperformant and has insurmountable challenges) is a self-fulfilling prophecy, and will remain so unless we invest in enough effort to disprove it.
+
+4) A fair lot of cryptographic projects in recent years has been specifically towards *cryptocurrencies*, rather than *cryptography in general*, and their results are at large fairly useless if they cannot be used outside of their ecosystems.
+
+All of these reasons have contributed to the stagnant state of Haskell cryptography, and have make it an unfavorable environment for developers of cryptosystems and distributed systems. However, none of these are actual blockers, and each of these things can be corrected with effort. There is much that can be done to improve the Haskell cryptography ecosystem, ranging from better low-level support, to higher-level abstractions:
+
+- Up-to-date algorithms
+- Hierarchy of typeclasses
+- Monadic interfaces
+- Safety abstractions
+- Type-level composition
```

Comment:
How is type-level composition something that's useful for cryptography? Similarly, why is a monadic interface better than, say, an Applicative one or just a pure function? My mental model of the semantics of a typical cipher is that they take a plaintext, a key, and a nonce, and then spit out a ciphertext - this is a pure function! Why is a monad useful here?

Similarly, I'd typically consider a hierarchy of type classes to be a bad thing, unless it's needed to solve a particular problem. It's nicer to have simpler APIs, isn't it?

Reply:
I'll be eliding some of this (type-level stuff) from the proposal, and Ill clear this up a bit. It is relevant for more abstract stuff down the line, but after a bunch of comments about the more abstract libraries, I decided it was distracting / out of scope of this proposal. Type-level stuff will be useful later for composing cryptographic primitives  but for wrapping Botan it is unnecessary.

Regarding monadic interfaces:

> Similarly, why is a monadic interface better

I sort of envision a `Crypto` monad that ensures strict IO, deep evaluation, memory safety, and prompt cleanup - a sort of cryptographically-oriented ST or STM, and its hard to see that working with just Applicative. A pure implementation should be used as the foundation if at all possible, but manually passing around random generators or outputs is a big potential error surface, and monads as a design pattern help minimize that.

> It's nicer to have simpler APIs, isn't it?

Absolutely, but not every library exposes pure APIs for their cryptographic operations. Even in Botan, there is some dissonance regarding this. The intent here isn't for us to force users to use a specific (abstract) interface, it is to have one already implemented for users when they come across an implementation that itself forces users to use a specific interface. I refer back to:

```haskell
deterministicFoo :: Int -> a -> b
randomFoo :: (Random m) => a -> m b
pseudoRandomFoo :: Pseudorandom g => g -> a -> (g, b)
impureFoo :: a -> IO b
```

Sometimes `deterministicFoo` is not exported / exposed / available - we might only get access to one of these functions, which leaves us unable to implement some or all of the others.

Admittedly there is a bit of tension between 'ideal' cryptography, the actual implementations available to us, and our need to perform cryptography safely.